### PR TITLE
Handle gevent exception in gevent poller

### PR DESCRIPTION
**Pull request by beandrad**
Currently the gevent poller ignores exceptions raised by
`gevent.wait()`, which causes greenlets to be unkilable while waiting.

This change handles exceptions raised while waiting in the gevent
poller, cancels the gRPC call and propagates the error back to the
application.

Co-authored-by: Kostis Lolos <klolos@arrikto.com>
See https://github.com/grpc/grpc/pull/19329
Fixes: #15880

@markdroth

### CODE REVIEW

**Comment by pranavgupta1234** [comment]
Would be nice to have this merged

File: tools/run_tests/run_tests.py [hunk 1]
```diff
@@ -622,13 +622,13 @@ class PythonConfig(
 class PythonLanguage(object):
 
     _TEST_SPECS_FILE = {
-        'native': 'src/python/grpcio_tests/tests/tests.json',
-        'gevent': 'src/python/grpcio_tests/tests/tests.json',
-        'asyncio': 'src/python/grpcio_tests/tests_aio/tests.json',
+        'native': ['src/python/grpcio_tests/tests/tests.json'],
+        'gevent': ['src/python/grpcio_tests/tests/tests.json', 'src/python/grpcio_tests/tests_gevent/tests.json',],
```
Comment (lidizheng) [inline tools/run_tests/run_tests.py:626]: Line 784: Let's disable Python27 gevent tests for Linux and macOS. You can do so by not returning 27's Python config.

File: src/python/grpcio_tests/tests_gevent/unit/_test_server.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+UNARY_CALL_WITH_SLEEP_VALUE = 0.2
+
+class TestServiceServicer(test_pb2_grpc.TestServiceServicer):
+    
+    def UnaryCallWithSleep(self, unused_request, unused_context):
+        gevent.sleep(UNARY_CALL_WITH_SLEEP_VALUE)
+        return messages_pb2.SimpleResponse()
+
+def start_test_server(port=0):
```
Comment (lidizheng) [inline src/python/grpcio_tests/tests_gevent/unit/_test_server.py:28]: Type annotations?

File: src/python/grpcio_tests/tests_gevent/unit/_test_server.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
```
Comment (lidizheng) [inline src/python/grpcio_tests/tests_gevent/unit/_test_server.py:29]: Let's use the default value of `max_workers`, it will be cores * 5.

File: src/python/grpcio_tests/tests_gevent/unit/_test_server.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    servicer = TestServiceServicer()
+    test_pb2_grpc.add_TestServiceServicer_to_server(TestServiceServicer(), server)
+
+    server.add_generic_rpc_handlers((_create_extra_generic_handler(servicer),))
+    port = server.add_insecure_port('[::]:%d' % port)
+    server.start()
+    return 'localhost:%d' % port, server
+
+def _create_extra_generic_handler(servicer: TestServiceServicer):
```
Comment (lidizheng) [inline src/python/grpcio_tests/tests_gevent/unit/_test_server.py:38]: Returning type?

File: src/python/grpcio/grpc/_cython/_cygrpc/grpc_gevent.pyx.pxi [hunk 5]
```diff
@@ -348,12 +352,25 @@ cdef void destroy_loop() with gil:
...
 
+cdef grpc_error* run_loop(size_t timeout_ms) with gil:
+  try:
+    _run_loop(timeout_ms)
+    return grpc_error_none()
+  except BaseException:
+    exc_info = sys.exc_info()
+    # Avoid running any Python code after setting the exception
+    cpython.PyErr_SetObject(exc_info[0], exc_info[1])
+    return grpc_error_cancelled()
```
Comment (lidizheng) [inline src/python/grpcio/grpc/_cython/_cygrpc/grpc_gevent.pyx.pxi:371]: Nice workaround!

File: src/python/grpcio/grpc/_cython/_cygrpc/grpc_gevent.pyx.pxi [hunk 6]
```diff
@@ -15,13 +15,17 @@
...
+import sys
 gevent_g = None
 gevent_socket = None
 gevent_hub = None
 gevent_event = None
 g_event = None
 g_pool = None
 
+cdef grpc_error* grpc_error_cancelled():
+  return <grpc_error*>4
```
Comment (lidizheng) [inline src/python/grpcio/grpc/_cython/_cygrpc/grpc_gevent.pyx.pxi:27]: The ground truth is at "src/core/lib/iomgr/error.h". We should import it via `grpc.pxi`. It helps the code health.

File: src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pyx.pxi [hunk 7]
```diff
@@ -12,10 +12,23 @@
...
 cdef int _INTERRUPT_CHECK_PERIOD_MS = 200
 
 
+cdef grpc_event _grpc_completion_queue_next(grpc_completion_queue *cq,
+                                            gpr_timespec deadline,
+                                            void *reserved) except *:
+    # Wrap the call to the C++ core in an 'except *' helper to propagate any
+    # errors raised by the gevent poller callback
+    with nogil:
+      return grpc_completion_queue_next(cq, deadline, reserved)
```
Comment (lidizheng) [inline src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pyx.pxi:29]: Do you have the performance impact percentage of this change? Our benchmark framework is a bit hard to use. I would recommend writing a simple client and server to measure it yourself.

**Review by lidizheng — COMMENTED** [review]
Sorry I didn't get to this PR quicker. Thanks for the reminder. This PR changes the interface of custom IO manager, I was skeptical if Node will be happy about it. Turns out the tests looks good (except Python27).

This PR is definitely a significant improvement for gevent integration. Thank you for digging deep into the gRPC repo and writing this PR.

File: src/core/lib/iomgr/pollset_custom.h [hunk 8]
```diff
@@ -19,13 +19,14 @@
...
 
+#include "src/core/lib/iomgr/error.h"
 #include <grpc/support/port_platform.h>
 
 #include <stddef.h>
 
 typedef struct grpc_custom_poller_vtable {
   void (*init)();
-  void (*poll)(size_t timeout_ms);
+  grpc_error* (*poll)(size_t timeout_ms);
```
Comment (gnossen) [inline src/core/lib/iomgr/pollset_custom.h:29]: @lidizheng As a change to the custom iomgr interface, do you know if this will impact Objective C?

File: src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pyx.pxi [hunk 9]
```diff
@@ -111,8 +126,8 @@ cdef class CompletionQueue:
         grpc_completion_queue_shutdown(self.c_completion_queue)
       # Pump the queue (All outstanding calls should have been cancelled)
       while not self.is_shutdown:
-        event = grpc_completion_queue_next(
+        event = _grpc_completion_queue_next(
             self.c_completion_queue, c_deadline, NULL)
         self._interpret_event(event)
       grpc_completion_queue_destroy(self.c_completion_queue)
-    grpc_shutdown()
+    grpc_shutdown_blocking()
```
Comment (gnossen) [inline src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pyx.pxi:133]: Can you please explain the change from an asynchronous shutdown to a synchronous one?

File: src/core/lib/iomgr/pollset_custom.h (hunk 8 above)
Comment (lidizheng) [reply src/core/lib/iomgr/pollset_custom.h:29]: ObjC's cfstream integration is several level higher at the endpoint interface. This PR won't impact ObjC, AFAIK.

File: tools/run_tests/run_tests.py [hunk 10]
```diff
@@ -789,7 +794,7 @@ def _get_pythons(self, args):
                 else:
                     return (python38_config,)
             else:
-                if args.iomgr_platform == 'asyncio':
+                if args.iomgr_platform in ('asyncio', 'gevent'):
```
Comment (beandrad) [inline tools/run_tests/run_tests.py:797]: @lidizheng is this the right way of disabling the tests for py27?

**Comment by beandrad** [comment]
I'm no longer able to access the results from the pipeline, but yesterday I saw that it was failing when running `tools/run_tests/sanity/check_port_platform.py`. Not sure how to fix that...
image

**Comment by lidizheng** [comment]
@beandrad We enabled the squash before merge feature. So, you don't need to force-push. It helps to locate test results and review.

File: src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pyx.pxi (hunk 7 above)
Comment (beandrad) [reply src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pyx.pxi:29]: I created a very simple [test](https://github.com/beandrad/grpc_perf_test); not sure if that's what you are referring to.

File: src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pyx.pxi (hunk 9 above)
Comment (beandrad) [reply src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pyx.pxi]: I've changed it back, no need for the change.

**Comment by beandrad** [comment]
@lidizheng not sure how to fix: `py37_native.test.unit._contextvars_propagation_test.ContextVarsPropagationTest` and `py38_native.test.unit._contextvars_propagation_test.ContextVarsPropagationTest`; could I disable them in https://github-beandrad/grpc/grpc/blob/0c898cda68cf32438a895b297d6d97a3f01f843e/src/python/grpcio_tests/commands.py#L175?

**Comment by gnossen** [comment]
@beandrad That is a [recently introduced high percentage flake on master](https://github.com/grpc/grpc/issues/26322) that I'm currently working on at a high priority. You can ignore it.

File: src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pyx.pxi [hunk 11]
```diff
@@ -12,10 +12,23 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+cimport cpython
+
+import threading
+import time
```
Comment (lidizheng) [inline src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pyx.pxi:18]: Unused imports?

File: src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pyx.pxi (hunk 7 above)
Comment (lidizheng) [reply src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pyx.pxi:29]: I played around the client. The QPS is much lower than my expectation, it should be 5k-7k for unary, 1 thread, gevent. Anyway, the old/new result demonstrated a 4% QPS regression (753.2 to 723.3).

Can you explain more about why we need `except *` to catch the exception, instead of using `PyErr_CheckSignals`?

File: src/python/grpcio_tests/tests_gevent/unit/close_channel_test.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+        self.assertFalse(self._unhandled_greenlet_exit,
+                         "Unhandled GreenletExit")
+
+    def _run_client(self):
+        UnaryCallWithSleep = self._channel.unary_unary(
+            _UNARY_CALL_METHOD_WITH_SLEEP,
+            request_serializer=messages_pb2.SimpleRequest.SerializeToString,
+            response_deserializer=messages_pb2.SimpleResponse.FromString,
+        )
+        while not self._channel_closed:
```
Comment (lidizheng) [inline src/python/grpcio_tests/tests_gevent/unit/close_channel_test.py:77]: If the test needs the RPC to be running while killing the greenlet, you may need some gevent synchronization here and the `_channel_closed` variable won't be needed.

Comment (beandrad) [reply src/python/grpcio_tests/tests_gevent/unit/close_channel_test.py:77]: `_channel_closed` is just used so that requests are not sent, but tbh it's not needed (or even sending the request). The purpose of this test is to check whether the `gevent` exception (`GreenletExit`) that kills the greenlet is handled in the event poller.

Comment (lidizheng) [reply src/python/grpcio_tests/tests_gevent/unit/close_channel_test.py:77]: Sorry for the nitpicking: Sleep + bool synchronization can be confusing in the long term. If possible, can we switch to a more deterministic logic and remove the sleep statements?

E.g., in `test_graceful_close` and `test_graceful_close_in_greenlet`, we expect the RPC to finish, we can use a simple unary call without sleep instead. For `test_ungraceful_close_in_greenlet`, we don't expect the RPC is finish, so the server method can sleep forever.

File: src/python/grpcio_tests/tests_gevent/unit/close_channel_test.py [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+import grpc
+import gevent
+import sys
+from gevent.pool import Group
+from tests_gevent.unit._test_server import start_test_server, UNARY_CALL_WITH_SLEEP_VALUE
+
+_UNARY_CALL_METHOD_WITH_SLEEP = '/grpc.testing.TestService/UnaryCallWithSleep'
+
+
+class CloseChannelTest(unittest.TestCase):
```
Comment (lidizheng) [inline src/python/grpcio_tests/tests_gevent/unit/close_channel_test.py:27]: This is a great test that checks the behavior around `GreenletExit`, which is a special exception. Can we try to inject a normal Python exception, and ensure it will populate to application?

Comment (beandrad) [reply src/python/grpcio_tests/tests_gevent/unit/close_channel_test.py:26]: I added `test_kill_greenlet_with_generic_exception`; it kills the greenlet with `Exception` instead of using the default `GreenletExit` and checks whether the greenlet returns the corresponding exception.

File: src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pyx.pxi (hunk 7 above)
Comment (beandrad) [reply src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pyx.pxi]: Otherwise the loop in [`_next()`](https://github.com/grpc/grpc/blob/99ac3cb03340a5dd4407bd696d6d5176b1d7eac7/src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pyx.pxi#L39) never breaks. Another solution is https://github.com/grpc/grpc/pull/26058/commits/b82a655e97af08120627f950cbed49c2449b239b, but I'm not sure if it'd have any side effect.

File: src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pyx.pxi [hunk 14]
```diff
@@ -12,6 +12,8 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+cimport cpython
+
```
Comment (lidizheng) [inline src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pyx.pxi:16]: Is this change necessary?

**Review by lidizheng — APPROVED** [review]
LGTM with a nit. Thanks for the update. @gnossen PTAL.

**Review by gnossen — APPROVED** [review]
Thanks for the contribution! :+1:

**Comment by beandrad** [comment]
`testHijackingInterceptor ` seems to be failing, could it be related to https://github.com/grpc/grpc/issues/25765?
